File: solutions/718_maximum_length_of_repeated_subarray.cpp
#include "aux.cpp"
/**
Given two integer arrays A and B, return the maximum length of an subarray that
appears in both arrays. Example 1: Input: A: [1,2,3,2,1] B: [3,2,1,4,7] Output:
3 Explanation: The repeated subarray with maximum length is [3, 2, 1]. Note: 1
<= len(A), len(B) <= 1000 0 <= A[i], B[i] < 100

 https://leetcode.com/problems/maximum-length-of-repeated-subarray/description/
 **/
static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();

class Solution {
public:
  int findLength(vector<int> &a, vector<int> &b) {
    vector<vector<int>> dp(a.size() + 1, vector<int>(b.size() + 1));
    int res = 0;
    for (int i = 0; i <= a.size(); i++)
      for (int j = 0; j <= b.size(); j++)
        if (i == 0 || j == 0)
          dp[i][j] = 0;
        else if (a[i - 1] == b[j - 1]) {
          dp[i][j] = 1 + dp[i - 1][j - 1];
          res = max(dp[i][j], res);
        }
    // say(dp);
    return res;
  }
};

int main() {
  Solution s;
  vector<int> a = {1, 2, 3, 2, 1};
  vector<int> b = {3, 2, 2};
  say(s.findLength(a, b));
}

File: python_solutions/718.maximum-length-of-repeated-subarray.py
#
# @lc app=leetcode id=718 lang=python3
#
# [718] Maximum Length of Repeated Subarray
#
# https://leetcode.com/problems/maximum-length-of-repeated-subarray/description/
#
# algorithms
# Medium (46.27%)
# Total Accepted:    35.1K
# Total Submissions: 75.8K
# Testcase Example:  '[1,2,3,2,1]\n[3,2,1,4,7]'
#
# Given two integer arrays A and B, return the maximum length of an subarray
# that appears in both arrays.
#
# Example 1:
#
#
# Input:
# A: [1,2,3,2,1]
# B: [3,2,1,4,7]
# Output: 3
# Explanation:
# The repeated subarray with maximum length is [3, 2, 1].
#
#
#
#
# Note:
#
#
# 1 <= len(A), len(B) <= 1000
# 0 <= A[i], B[i] < 100
#
#
#
#
#
class Solution:
    def findLength(self, A,B):
        def check(i):
            seen = set(A[x:x + i] for x in range(len(A) - i + 1))
            return any(B[x:x + i] in seen for x in range(len(B) - i + 1))
        
        l, r = 0, min(len(A), len(B))
        A = ''.join(map(chr, A))
        B = ''.join(map(chr, B))
        while l < r:
            mid = (l + r) // 2
            if check(mid):
                l = mid + (l == mid)
            else:
                r = mid - 1
        return l if check(l) else l - 1

a, b = [1, 2, 3, 2, 1], [3, 2, 1, 4, 7]
# a, b = [0, 1, 1, 1, 1], [1, 0, 1, 0, 1]
s = Solution()
print(s.findLength(a, b))
